File: sdk/kotlin/src/main/kotlin/com/zcashme/zns/Verify.kt
package com.zcashme.zns

import java.security.KeyFactory
import java.security.Signature
import java.util.Base64

fun verifySignature(payload: String, signatureB64: String, pubkeyB64: String): Boolean {
    return try {
        val sigBytes = Base64.getDecoder().decode(signatureB64)
        val pkBytes = Base64.getDecoder().decode(pubkeyB64)
        if (sigBytes.size != 64 || pkBytes.size != 32) return false
        val pkSpec = java.security.spec.EdECPublicKeySpec(
            java.security.spec.NamedParameterSpec("Ed25519"),
            java.security.spec.EdECPoint(false, decodeUint256(pkBytes))
        )
        val pk = KeyFactory.getInstance("Ed25519").generatePublic(pkSpec)
        val verifier = Signature.getInstance("Ed25519")
        verifier.initVerify(pk)
        verifier.update(payload.toByteArray(Charsets.UTF_8))
        verifier.verify(sigBytes)
    } catch (_: Exception) {
        false
    }
}

private fun decodeUint256(littleEndian: ByteArray): java.math.BigInteger {
    val reversed = littleEndian.copyOf().also { it.reverse() }
    return java.math.BigInteger(1, reversed)
}

fun registrationPayload(reg: Registration): String = when (reg.lastAction) {
    "CLAIM" -> "CLAIM:${reg.name}:${reg.address}"
    "BUY" -> "BUY:${reg.name}:${reg.address}"
    "UPDATE" -> "UPDATE:${reg.name}:${reg.address}:${reg.nonce}"
    "DELIST" -> "DELIST:${reg.name}:${reg.nonce}"
    "RELEASE" -> "RELEASE:${reg.name}:${reg.nonce}"
    else -> ""
}

fun listingPayload(listing: Listing): String =
    "LIST:${listing.name}:${listing.price}:${listing.nonce}"

fun verifyRegistration(reg: Registration, adminPubkey: String): Boolean {
    val sig = reg.signature ?: return false
    val payload = registrationPayload(reg)
    if (payload.isEmpty()) return false
    val pubkey = reg.pubkey ?: adminPubkey
    return verifySignature(payload, sig, pubkey)
}

fun verifyListing(listing: Listing, adminPubkey: String): Boolean {
    val payload = listingPayload(listing)
    return verifySignature(payload, listing.signature, adminPubkey)
}

File: sdk/python/zcashname/verify.py
from __future__ import annotations

import base64
from typing import TYPE_CHECKING

from nacl.signing import VerifyKey
from nacl.exceptions import BadSignatureError

if TYPE_CHECKING:
    from .types import Registration, Listing


def verify_signature(payload: str, signature_b64: str, pubkey_b64: str) -> bool:
    try:
        sig_bytes = base64.b64decode(signature_b64)
        pk_bytes = base64.b64decode(pubkey_b64)
        if len(sig_bytes) != 64 or len(pk_bytes) != 32:
            return False
        vk = VerifyKey(pk_bytes)
        vk.verify(payload.encode(), sig_bytes)
        return True
    except (BadSignatureError, Exception):
        return False


def registration_payload(reg: Registration) -> str:
    match reg.last_action:
        case "CLAIM":
            return f"CLAIM:{reg.name}:{reg.address}"
        case "BUY":
            return f"BUY:{reg.name}:{reg.address}"
        case "UPDATE":
            return f"UPDATE:{reg.name}:{reg.address}:{reg.nonce}"
        case "DELIST":
            return f"DELIST:{reg.name}:{reg.nonce}"
        case "RELEASE":
            return f"RELEASE:{reg.name}:{reg.nonce}"
        case _:
            return ""


def listing_payload(listing: Listing) -> str:
    return f"LIST:{listing.name}:{listing.price}:{listing.nonce}"


def verify_registration(reg: Registration, admin_pubkey: str) -> bool:
    if not reg.signature:
        return False
    payload = registration_payload(reg)
    if not payload:
        return False
    pubkey = reg.pubkey or admin_pubkey
    return verify_signature(payload, reg.signature, pubkey)


def verify_listing(listing: Listing, admin_pubkey: str) -> bool:
    payload = listing_payload(listing)
    return verify_signature(payload, listing.signature, admin_pubkey)